envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/JwtFilterFactory.kt

```diff
@@ -0,0 +1,103 @@
+package pl.allegro.tech.servicemesh.envoycontrol.snapshot.resource.listeners.filters
+
+import com.google.protobuf.Any
+import com.google.protobuf.Empty
+import com.google.protobuf.util.Durations
+import io.envoyproxy.envoy.config.core.v3.HttpUri
+import io.envoyproxy.envoy.config.route.v3.RouteMatch
+import io.envoyproxy.envoy.extensions.filters.http.jwt_authn.v3.JwtAuthentication
+import io.envoyproxy.envoy.extensions.filters.http.jwt_authn.v3.JwtProvider
+import io.envoyproxy.envoy.extensions.filters.http.jwt_authn.v3.JwtRequirement
+import io.envoyproxy.envoy.extensions.filters.http.jwt_authn.v3.RemoteJwks
+import io.envoyproxy.envoy.extensions.filters.http.jwt_authn.v3.RequirementRule
+import io.envoyproxy.envoy.extensions.filters.network.http_connection_manager.v3.HttpFilter
+import pl.allegro.tech.servicemesh.envoycontrol.groups.Group
+import pl.allegro.tech.servicemesh.envoycontrol.groups.IncomingEndpoint
+import pl.allegro.tech.servicemesh.envoycontrol.groups.OAuth
+import pl.allegro.tech.servicemesh.envoycontrol.snapshot.JwtFilterProperties
+import pl.allegro.tech.servicemesh.envoycontrol.snapshot.OAuthProvider
+
+class JwtFilterFactory(
+    private val properties: JwtFilterProperties
+) {
+
+    fun createJwtFilter(group: Group): HttpFilter? {
+        return if (shouldCreateFilter(group)) {
+            HttpFilter.newBuilder()
+                .setName("envoy.filters.http.jwt_authn")
+                .setTypedConfig(
+                    Any.pack(
+                        JwtAuthentication.newBuilder().putAllProviders(
+                            getOAuthProviders()
+                        ).addAllRules(
+                            createRules(group.proxySettings.incoming.endpoints)
+                        )
+                            .build()
+                    )
+                )
+                .build()
+        } else {
+            null
+        }
+    }
+
+    private fun shouldCreateFilter(group: Group): Boolean {
+        return group.proxySettings.incoming.endpoints.any { it.oauth != null } && properties.providers.isNotEmpty()
+    }
+
+    private fun getOAuthProviders(): Map<String, JwtProvider> =
```

Comment:
The same, we can make it at the beginning and won't need to create new instance each time